File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3027_find-the-number-of-ways-to-place-people-ii/solutions/java/solution.java
import java.util.*;

class Solution {
    public int numberOfPairs(int[][] points) {
        // Step 1: Place points into a set for fast lookup
        Set<String> pointSet = new HashSet<>();
        for (int[] p : points) {
            pointSet.add(p[0] + "," + p[1]);
        }

        // Step 2: Initialize count of valid pairs
        int count = 0;
        
        // Step 3: Check pairs
        for (int i = 0; i < points.length; i++) {
            for (int j = 0; j < points.length; j++) {
                if (i == j) continue; // skip self-comparison

                int x1 = points[i][0], y1 = points[i][1];
                int x2 = points[j][0], y2 = points[j][1];

                // Alice must be at the upper-left and Bob at the lower-right
                if (x1 <= x2 && y1 >= y2) {
                    boolean valid = true;
                    // Check all points for being inside or on the rectangle boundary
                    for (int k = 0; k < points.length; k++) {
                        if (k == i || k == j) continue;
                        int x = points[k][0], y = points[k][1];
                        if (x >= x1 && x <= x2 && y <= y1 && y >= y2) {
                            valid = false;
                            break;
                        }
                    }
                    if (valid) count++;
                }
            }
        }

        return count;
    }
}

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3027_find-the-number-of-ways-to-place-people-ii/solutions/python3/solution.py
from typing import List

class Solution:
    def numberOfPairs(self, points: List[List[int]]) -> int:
        # Converting points into a set for faster lookup
        point_set = set(tuple(point) for point in points)
        valid_pairs = 0
        
        # We need to check each point against every other to find suitable pairs
        for i in range(len(points)):
            x1, y1 = points[i]
            for j in range(len(points)):
                if i != j:
                    x2, y2 = points[j]
                    
                    # Alice should be at the upper left corner, and Bob at the lower right corner
                    if x1 <= x2 and y1 >= y2:
                        # We must ensure no other points lie on the rectangle edges formed by (x1, y1) and (x2, y2)
                        no_interior_points = True
                        for x, y in points:
                            # If it's neither Alice nor Bob and lies inside or on the boundary of the rectangle
                            if (x1 < x < x2 and y1 > y > y2) or (
                                (x == x1 or x == x2) and y1 >= y >= y2) or (
                                (y == y1 or y == y2) and x1 <= x <= x2):
                                no_interior_points = False
                                break
                        
                        if no_interior_points:
                            valid_pairs += 1
        
        return valid_pairs
